### PR TITLE
[SPARK-17244] Catalyst should not pushdown non-deterministic join conditions

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

Given that non-deterministic expressions can be stateful, pushing them down the query plan during the optimization phase can cause incorrect behavior. This patch fixes that issue by explicitly disabling that.
## How was this patch tested?

A new test in `FilterPushdownSuite` that checks catalyst behavior for both deterministic and non-deterministic join conditions.

### CODE REVIEW

**Comment by sameeragarwal** [comment]
cc @hvanhovell @gatorsmile

**Comment by sameeragarwal** [comment]
cc @brkyvz who found this bug

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1386,15 +1386,17 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
 object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
   /**
    * Splits join condition expressions into three categories based on the attributes required
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It sounds like this PR is targeting both join conditions and filter predicates. Could you update the code comment of this line and the title of the PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1386,15 +1386,17 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
...
+   * to evaluate them. Note that we explicitly exclude non-deterministic (i.e., stateful) condition
+   * expressions in canEvaluateInLeft or canEvaluateInRight to prevent pushing these predicates on
+   * either side of the join.
    *
    * @return (canEvaluateInLeft, canEvaluateInRight, haveToEvaluateInBoth)
    */
   private def split(condition: Seq[Expression], left: LogicalPlan, right: LogicalPlan) = {
     val (leftEvaluateCondition, rest) =
-        condition.partition(_.references subsetOf left.outputSet)
+        condition.partition(expr => expr.references.subsetOf(left.outputSet) && expr.deterministic)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Maybe we need to use `span` here. See another PR: https://github.com/apache/spark/pull/14012

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Good catch! Didn't realize that relative ordering of these expressions could become an issue.

**Comment by sameeragarwal** [comment]
Thanks @gatorsmile, I address your comments.

**Comment by gatorsmile** [comment]
Looks pretty good to me! 

Could you also add one more test case for testing the following scenario: 
```
only consider pushing down those expressions that precede the first non-deterministic expression in the condition.
```

I am just afraid the others might break it in the future. Thanks!

**Comment by sameeragarwal** [comment]
That's a good idea. More generally, I think the best way to fix these class of problems is to unify all predicate pushdown logic at a common place.

**Comment by gatorsmile** [comment]
In 2.0, we combined the PPD rules. Now, only two PPD rules left, `PushPredicateThroughJoin` and `PushDownPredicate`. IMO, the major reason why we did not combine them into the same one is because PPD for `Join` is a little bit special.

**Comment by gatorsmile** [comment]
LGTM pending testing

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 2.0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -1385,18 +1385,25 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
  */
 object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
   /**
-   * Splits join condition expressions into three categories based on the attributes required
-   * to evaluate them.
+   * Splits join condition expressions or filter predicates (on a given join's output) into three
+   * categories based on the attributes required to evaluate them. Note that we explicitly exclude
+   * on-deterministic (i.e., stateful) condition expressions in canEvaluateInLeft or
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: type: `non-deterministic`

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: good eye! can you please fold this change into one of your open PRs :) ?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: yea, I will :)

**Pull request closed without merging**